### PR TITLE
FSD50K Speech Model Fine-tuning Tutorial

### PR DESCRIPTION
**Pull request by FlorentMeyer**
## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## What does this PR do?

Add FSD50K Speech Model Fine-tuning Tutorial.

## Did you have fun?

A lot 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
hey @FlorentMeyer, mind check the file you uploaded, looks like it's too big and there might be some redundant stuff here. Might clean it up?

**Comment by Borda** [comment]
not sure what happen but GH does not want to show me the diff :/

**Comment by FlorentMeyer** [comment]
Good evening,

I should mention that the code in the converted notebook was exactly the same as in this [Colab notebook](https://colab.research.google.com/drive/1NddRCV1BtwgK6tvnylkLHY8d7t4OhAEw?usp=sharing) (having removed the `!pip install`s). I also kept the output, but reading other people's examples I suppose that the outputs printed inside the docs are the ones obtained by running the .py converted notebooks on your side.

My last commit therefore makes these changes to the linked Colab notebook:
- remove all cells outputs
- remove conditions on bash instructions (a single bash command inside an `if` was causing a syntax error)
- remove %% magic
- comment out the tensorboard cell (which was responsible for creating such a large file, I am sorry I hadn't checked it before) 

Changes to the .yaml file:
- add gdown as a requirement
- remove brackets around my name (didn't know they were special characters)

I'm just not sure whether the Pandas dataframes with the audio players will get rendered.

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+# %% [markdown] id="CI0JECKA9AnY"
+# # README
+#
+# MWE of fine-tuning a Transformer-based speech embedder (e.g. [wav2vec 2.0](https://arxiv.org/abs/2006.11477)) on a subset of FSD50K using `pytorch_lightning` and HuggingFace `transformers`.
+#
+# Please refer to this executable [Colab notebook](https://colab.research.google.com/drive/1NddRCV1BtwgK6tvnylkLHY8d7t4OhAEw?usp=sharing) importing the code from this repo as well as a 500-element subset of the original FSD50K dataset for a concrete train+test example.
+#
+# Note: intended as an editable incentive for jumping into FSD50K and the Pytorch-Lightning+HuggingFace framework, and as a showcase for an end-of-studies project -- choices have been made and some logic has been altered to (greatly) reduce the size of the original code.
+#
```
Comment (rohitgr7) [inline lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:23]:
```suggestion
# MWE of fine-tuning a Transformer-based speech embedder (e.g. [wav2vec 2.0](https://arxiv.org/abs/2006.11477)) on a subset of FSD50K using `pytorch_lightning` and HuggingFace `transformers`.
#
# Please refer to this executable [Colab notebook](https://colab.research.google.com/drive/1NddRCV1BtwgK6tvnylkLHY8d7t4OhAEw?usp=sharing) importing the code from this repo as well as a 500-element subset of the original FSD50K dataset for a concrete train+test example.
#
# Note: intended as an editable incentive for jumping into FSD50K and the Pytorch-Lightning+HuggingFace framework and as a showcase for an end-of-studies project -- choices have been made, and some logic has been altered to (significantly) reduce the size of the original code.
#

```

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+# %% [markdown] id="oqwX4z1RT8Uw"
+# ## Explore samples with highest prediction scores
+
+# %% [markdown] id="SYH5XclifZpJ"
+# Retrieve audio file paths from their IDs.
+#
+# For each of the samples on which a prediction was made, rank (for example the first 4, hence the `"logits_4_highest"` key) highest confidence logits with their corresponding class names.
+#
+# As an example for choosing some samples among the 100 in the test set, rank samples according to the highest logit they contain to inspect audios for which the model seemed confident -- as an alternative, one could also choose a handful of samples at random.
+
```
Comment (rohitgr7) [inline lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:260]:
```suggestion
# %% [markdown] id="oqwX4z1RT8Uw"
# ## Explore samples with the highest prediction scores

# %% [markdown] id="SYH5XclifZpJ"
# Retrieve audio file paths from their IDs.
#
# For each of the samples on which a prediction was made, rank (for example, the first 4, hence the `"logits_4_highest"` key) the highest confidence logits with their corresponding class names.
#
# As an example for choosing some samples among the 100 in the test set, rank samples according to the highest logit they contain to inspect audios for which the model seemed confident -- as an alternative, one could also choose a handful of samples at random.


```

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+# # Evaluate performance
+
+# %% [markdown] id="bPbyVevnZXTX"
+# ## Predict on test data
+
+# %% id="6VTZHQDw2y5p"
+preds = trainer.predict(ckpt_path="best", datamodule=datamodule)
+
+# %% id="FM0fc3UbSp-2"
+preds = gather_preds(preds)
```
[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:238]: It takes outputs from pl.trainer.predict() as a list of batches dicts to gather into a single dict.

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+# %% [markdown] id="cD556ynvAbYa"
+# ## Compute metrics
+
+# %% id="zlTooqqp8FWk"
+mAP_micro = average_precision_score(
```
[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: At the time I wrote the code, `torchmetrics.functional.average_precision`'s `target` took "integer labels" therefore not accepting multi-hot labels. Just let me check whether this was fixed and if I get the same results as with scikit-learn!

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+#
+# As an example for choosing some samples among the 100 in the test set, rank samples according to the highest logit they contain to inspect audios for which the model seemed confident -- as an alternative, one could also choose a handful of samples at random.
+
+# %% id="N3Dk23p_O-38"
+preds = get_preds_fpaths(preds)
+preds = sort_highest_logits(preds, fsd50k_datadict.token_to_name, num_classes=4)
+preds_max_logits_indices = get_preds_max_logits_indices(preds)
+
+# %% id="Vnyr_f7tmNYJ"
+inspect_data(
```
[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:270]: It displays a Pandas dataframe containing text data as well as audio players. It's used for inspecting a dataset/predictions with their audio, their labels, the predicted logits, etc. You can find an example here: https://colab.research.google.com/drive/1NddRCV1BtwgK6tvnylkLHY8d7t4OhAEw?usp=sharing#scrollTo=UwDU12Ewj-fA

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: OK, the new implementations of `multilabel_average_precision` give the same results as scikit-learn

[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: What do you mean? Something like this?

With `on_step=False, on_epoch=True` to only log the end of the epoch according to https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule:
> The above config for validation applies for test hooks as well.

```suggestion
# In __init__:
self.mAP = torchmetrics.classification.MultilabelAveragePrecision()

# In test_step:
self.mAP(preds, y)
self.log('mAP', self.mAP)
```

[the remark this answers is not in the document]
Comment (FlorentMeyer) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: Then the class version of torchmetrics should be prefered to functional I'd say?

[the remark this answers is not in the document]
Comment (rohitgr7) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: okay.. it's fine.. let's use functional metrics here since you already have all the targets and predictions.
modular metrics are useful, when you are aggregating the metrics, let's say on step-level

[the remark this answers is not in the document]
Comment (rohitgr7) [reply lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:244]: let's use that :)

**Comment by FlorentMeyer** [comment]
I see there also are problems with:
1) the git+<my_repo> in requirement.txt
2) the building of the docs saying some cells are missing IDs but I used jupytext as requested

**Comment by FlorentMeyer** [comment]
OK I also saw that there were bizarre things happening in the notebook, looks like the pre-commit hooks are moving stuff around causing duplication every time I pull them into my own code before being able to push again ([example](https://github.com/Lightning-AI/tutorials/pull/201/commits/344c7aba22d260338baf341a209d6bdbc407248d#diff-5a1e1961c94d46c19724f41e83a47f3c5103e098c40a7125c2e83d40935c0be6R65-R75)) and it's easy to miss things when reading a notebook as a .py file.

Anyway I read the whole file carefully and this should be fixed now. Also all cells have an ID so I'm not sure where the "cells are missing IDs" error comes from :/

**Comment by FlorentMeyer** [comment]
Small up!

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+#     formats: ipynb,py:percent
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.14.1
+#   kernelspec:
+#     display_name: Python 3
+#     name: python3
+# ---
```
Comment (Borda) [inline lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:13]:
```suggestion
```

File: lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+# %% [markdown] id="CI0JECKA9AnY"
```
Comment (Borda) [inline lightning_examples/fsd50K-speech-model-finetuning/fsd50K_speech_model_finetuning.py:15]: lets remove all the ID

**Comment by Borda** [comment]
seem the repo https://github.com/FlorentMeyer/fsd50k_speech_model_finetuning and this example is silent since 2022, so let's close it for now, but feel free to reopen any time :flamingo:
